app/backend/src/core/controllers/appraiser/handlers/TokenTransfers.ts

```diff
@@ -0,0 +1,157 @@
+    const transferEvents: TransferLogEvent[] = this.extractTransferEvents(
+      receipt.logs,
+    );
+
+    const hasReciprocal = this.hasReciprocalTransfers(transferEvents);
+    return hasReciprocal
+      ? await this.handleSwapLikeTransfers(transferEvents, timestamp)
+      : await this.handleDistributionLikeTransfers(transferEvents, timestamp);
+  }
+
+  private extractTransferEvents(logs: Log[]): TransferLogEvent[] {
+    const transferEvents: TransferLogEvent[] = [];
+
+    const transferEventSigHash = ethers.id("Transfer(address,address,uint256)");
+
+    for (const log of logs) {
+      if (log.topics[0] === transferEventSigHash) {
+        const contractAddress = log.address;
+        const fromAddress = ethers.getAddress(`0x${log.topics[1].slice(26)}`);
+        const toAddress = ethers.getAddress(`0x${log.topics[2].slice(26)}`);
+        const rawAmount = log.data === "0x" ? BigInt(0) : BigInt(log.data);
+        transferEvents.push({
+          fromAddress,
+          toAddress,
+          contractAddress,
+          rawAmount,
+        });
+      }
+    }
+    return transferEvents;
+  }
+
+  private async handleSwapLikeTransfers(
+    transferEvents: TransferLogEvent[],
+    timestamp: UnixTime,
+  ): Promise<AppraisalSummary[]> {
+    const pricedTransferLogEvents: PricedTransferLogEvent[] =
+      await this.priceTransferLogEvents(transferEvents, timestamp);
+    const maxTransferEvent = pricedTransferLogEvents.reduce(
+      (max, event) =>
+        (max.usdValue || 0) < (event.usdValue || 0) ? event : max,
+      pricedTransferLogEvents[0],
+    );
+
+    return [
+      {
+        contractAddress: maxTransferEvent.contractAddress,
+        rawAmount: maxTransferEvent.rawAmount,
+        adjustedAmount: maxTransferEvent.adjustedAmount,
+        usdValue: maxTransferEvent.usdValue,
+      },
+    ];
+  }
+
+  private async handleDistributionLikeTransfers(
+    transferEvents: TransferLogEvent[],
+    timestamp: UnixTime,
+  ): Promise<AppraisalSummary[]> {
+    const pricedTransferLogEvents: PricedTransferLogEvent[] =
+      await this.priceTransferLogEvents(transferEvents, timestamp);
+
+    return pricedTransferLogEvents.map((event) => ({
+      contractAddress: event.contractAddress,
+      rawAmount: event.rawAmount,
+      adjustedAmount: event.adjustedAmount,
+      usdValue: event.usdValue,
+    }));
+  }
+
+  private hasReciprocalTransfers(transferEvents: TransferLogEvent[]): boolean {
+    const senders = new Set();
+    const receivers = new Set();
+
+    for (const event of transferEvents) {
+      senders.add(event.fromAddress);
+      receivers.add(event.toAddress);
+    }
+
+    return [...senders].some((sender) => receivers.has(sender));
+  }
+
+  private async priceTransferLogEvents(
+    transferEvents: TransferLogEvent[],
+    timestamp: UnixTime,
+  ): Promise<PricedTransferLogEvent[]> {
+    const pricedTransferEventPromises: Promise<PricedTransferLogEvent>[] =
+      transferEvents.map(async (event) => {
```

Comment:
Fetching latest prices for all the tokens needed in a single DB query as a dictionary and manipulating with it later will probably be more efficient.

Reply:
Agreed to address later